pkg/pipeline/encode/encode_prom.go

```diff
@@ -28,49 +29,56 @@ import (
 	"net/http"
 	"os"
 	"strconv"
+	"sync"
+	"time"
 )
 
-// TODO when do we clean up old entries?
+const defaultExpiryTime = 120
 
-type gaugeInfo struct {
-	input     string
-	tags      []string
-	promGauge *prometheus.GaugeVec
-}
-
-type counterInfo struct {
-	input     string
-	tags      []string
+type genericMetricInfo struct {
+	input       string
+	tags        []string
+	metricType  string
+	promGauge   *prometheus.GaugeVec
 	promCounter *prometheus.CounterVec
+	promHist    *prometheus.HistogramVec
 }
 
-type histInfo struct {
-	input    string
-	tags     []string
-	promHist *prometheus.HistogramVec
+type entrySignature struct {
+	Name   string
+	Labels map[string]string
 }
 
-type counterEntryInfo struct {
-	counterName  string
-	counterValue float64
-	labels     map[string]string
+type entryInfo struct {
+	eInfo entrySignature
+	value float64
 }
 
-type gaugeEntryInfo struct {
-	gaugeName  string
-	gaugeValue float64
-	labels     map[string]string
+type metricCacheEntry struct {
+	label       prometheus.Labels
```

Comment:
a lot of overlapping fields with the  `genericMetricInfo struct`. Maybe you can embed this struct in the cache struct? 

Look here: https://eli.thegreenplace.net/2020/embedding-in-go-part-1-structs-in-structs/

Or maybe have a dedicated field pointing to a  `genericMetricInfo struct`

Reply:
Factored out the common fields into PromMetric struct.